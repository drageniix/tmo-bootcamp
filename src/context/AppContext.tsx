import React, { createContext } from "react";
import { BasicIngredient, Recipe } from "../api/recipe";
import { useRecipeSearch } from "../hooks/context";

export type AppContextType = {
  onSearchIngredients?: (input: string) => Promise<void>;
  onAddIngredient?: (chosenIngredient: BasicIngredient) => Promise<void>;
  onRemoveIngredient?: (
    id: number
  ) => (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => Promise<void>;
  ingredients: BasicIngredient[];
  searchedIngredients: BasicIngredient[];
  recipes: Recipe[];
};

export const defaultContext: AppContextType = {
  ingredients: [],
  recipes: [],
  searchedIngredients: [],
};

export const AppContext = createContext(defaultContext);

export const AppProvider = ({
  children,
}: {
  children: JSX.Element;
}): JSX.Element => {
  const appProps = useRecipeSearch();

  return <AppContext.Provider value={appProps}>{children}</AppContext.Provider>;
};
