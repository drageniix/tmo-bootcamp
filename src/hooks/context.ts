import { useState } from "react";
import {
  BasicIngredient,
  searchIngredients,
  searchRecipes,
} from "../api/recipe";
import { AppContextType, defaultContext } from "../context/AppContext";

export const useRecipeSearch = (): AppContextType => {
  const [recipes, setRecipes] = useState(defaultContext.recipes);
  const [ingredients, setIngredients] = useState(defaultContext.ingredients);
  const [searchedIngredients, setSearchedIngredients] = useState(
    defaultContext.searchedIngredients
  );

  const onSearchIngredients = async (input: string) => {
    const ingredients = await searchIngredients(input);
    setSearchedIngredients(ingredients);
  };

  const onAddIngredient = async (chosenIngredient: BasicIngredient) => {
    if (!ingredients.includes(chosenIngredient)) {
      const updatedIngredients = [...ingredients, chosenIngredient];
      const recipes = await searchRecipes(
        updatedIngredients.map((ingredient) => ingredient.name)
      );

      setIngredients(updatedIngredients);
      setRecipes(recipes);
    }
  };

  const onRemoveIngredient = (id: number) => async (
    e: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => {
    e.preventDefault();
    const updatedIngredients = ingredients.filter(
      (ingredient) => ingredient.id !== id
    );
    const recipes = await searchRecipes(
      updatedIngredients.map((ingredient) => ingredient.name)
    );

    setIngredients(updatedIngredients);
    setRecipes(recipes);
  };

  return {
    onSearchIngredients,
    onAddIngredient,
    onRemoveIngredient,
    searchedIngredients,
    ingredients,
    recipes,
  };
};
