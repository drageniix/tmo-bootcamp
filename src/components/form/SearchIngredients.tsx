import React, { useContext } from "react";
import { AppContext } from "../../context/AppContext";
import { useComboboxIngredients } from "../../hooks/ingredients";
import IngredientOptions from "./IngredientOptions";

export const SearchIngredients = (): React.ReactElement => {
  const { onAddIngredient } = useContext(AppContext);
  const combobox = useComboboxIngredients();

  return (
    <section className="ingredientSearch" data-testid="form">
      <label {...combobox.getLabelProps()} className="ingredientSearch__label">
        Choose an ingredient:
      </label>
      <div
        {...combobox.getComboboxProps()}
        className="ingredientSearch__actions"
      >
        <input
          {...combobox.getInputProps()}
          className="ingredientSearch__actions--input"
        />
        <button
          type="button"
          className="ingredientSearch__actions--button"
          onClick={() => onAddIngredient(combobox.selectedItem)}
          aria-label={"add ingredient"}
        >
          +
        </button>
      </div>
      <IngredientOptions combobox={combobox} />
    </section>
  );
};

export default SearchIngredients;
