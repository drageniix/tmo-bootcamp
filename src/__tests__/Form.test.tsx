import React from "react";
import { cleanup, render, fireEvent } from "@testing-library/react";
import { AppContext, AppContextType } from "../context/AppContext";
import SearchIngredients from "../components/form/SearchIngredients";
import mockIngredients from "./mocks/ingredients.json";

const renderForm = (context?: AppContextType) =>
  render(
    <AppContext.Provider value={context}>
      <SearchIngredients />
    </AppContext.Provider>
  );

describe("should test form functionality", () => {
  afterEach(cleanup);

  it("should match form snapshot", () => {
    expect.assertions(1);

    const { getByTestId } = renderForm({
      ingredients: [],
      recipes: [],
      searchedIngredients: mockIngredients.results,
    });

    const formComponent = getByTestId("form") as HTMLInputElement;
    expect(formComponent).toMatchSnapshot();
  });

  it("should search ingredients when input is changed", () => {
    expect.assertions(1);

    const mockEvent = {
      target: { value: "banana" },
    };

    const searchIngredientsMock = jest.fn();

    const { getByRole } = renderForm({
      onSearchIngredients: searchIngredientsMock,
      ingredients: [],
      recipes: [],
      searchedIngredients: [],
    });

    const inputComponent = getByRole("textbox") as HTMLInputElement;
    fireEvent.change(inputComponent, mockEvent);

    expect(searchIngredientsMock).toHaveBeenCalledTimes(1);
  });

  it("should search recipes when add button is clicked", () => {
    expect.assertions(1);
    const mockedAddIngredient = jest.fn();

    const { getByRole } = renderForm({
      onAddIngredient: mockedAddIngredient,
      ingredients: [],
      recipes: [],
      searchedIngredients: [],
    });

    const buttonComponent = getByRole("button") as HTMLInputElement;
    fireEvent.click(buttonComponent);

    expect(mockedAddIngredient).toHaveBeenCalledTimes(1);
  });
});
