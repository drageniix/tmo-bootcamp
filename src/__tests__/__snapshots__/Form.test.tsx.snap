// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test form functionality should match form snapshot 1`] = `
<section
  class="ingredientSearch"
  data-testid="form"
>
  <label
    class="ingredientSearch__label"
    for="downshift-0-input"
    id="downshift-1-label"
  >
    Choose an ingredient:
  </label>
  <div
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-owns="downshift-1-menu"
    class="ingredientSearch__actions"
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      aria-controls="downshift-1-menu"
      aria-labelledby="downshift-1-label"
      autocomplete="off"
      class="ingredientSearch__actions--input"
      id="downshift-0-input"
      value=""
    />
    <button
      aria-label="add ingredient"
      class="ingredientSearch__actions--button"
      type="button"
    >
      +
    </button>
  </div>
  <ul
    aria-labelledby="downshift-1-label"
    class="ingredientSearch__list"
    id="downshift-1-menu"
    role="listbox"
  />
</section>
`;
